File: 0049-Group-Anagrams/lc0049.py
# Python solution.

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        hash = collections.defaultdict(list)
        for s in strs:
            buckets = [0] * 26
            for c in s:
                buckets[ord(c) - 97] += 1
            hash[tuple(buckets)].append(s)
        return list(hash.values())

File: 0049-Group-Anagrams/lc0049.rs
// Rust solution.

use std::collections::HashMap;

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let countsort = |s: &str| {
            s.bytes().fold([0;26], |mut buckets, c| {
                buckets[(c - b'a') as usize] += 1_u32;
                buckets
            })
        };

        let map = strs.into_iter().fold(HashMap::new(), |mut hash, s| {
            hash.entry(countsort(&s)).or_insert(Vec::new()).push(s);
            hash
        });

        map.into_iter().map(|(_, v)| v).collect::<Vec<_>>()
    }
}
